src/librustc_middle/query/mod.rs

```diff
@@ -582,6 +582,14 @@ rustc_queries! {
             desc { "destructure constant" }
         }
 
+        /// Dereference a constant reference or raw pointer and turn the result into a constant
+        /// again.
+        query deref_const(
```

Comment:
I wonder if this should just be `destructure_const`, maybe that could return an `enum`?

Reply:
All the uses sites would then have to handle either case, which for convenience would mean we'd add `unwrap` functions to it, which imo just increases the amount of boilerplate needed when calling this function. I could make these two functions inherent methods on `TyCtxt` and have them invoke a query that returns an enum though if you like.

Reply:
> All the uses sites would then have to handle either case,

Fair enough.